src/main/java/me/magnet/consultant/RoutingStrategies.java

```diff
@@ -0,0 +1,110 @@
+package me.magnet.consultant;
+
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+public class RoutingStrategies {
+
+	public static final RoutingStrategy RANDOMIZED_WEIGHTED_DISTANCE = new RoutingStrategy() {
+
+		private final Random random = new Random();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> instances = serviceLocator.listInstances(serviceName);
+			if (instances.size() <= 1) {
+				// No routing to do.
+				return instances;
+			}
+
+			// Pick a random number between 0-1.
+			double chance = random.nextDouble();
+
+			/*
+			 * Pick the index of the first instance to try based on this chance.
+			 * If the chance was < 0.5 then use the instance on index 0.
+			 * If the chance was >= 0.5 and < 0.75 then use the instance on index 1.
+			 * If the chance was >= 0.75 and < 0.875 then use the instance on index 2.
+			 * If the chance was >= 0.875 and < 0.9375 then use the instance on index 3.
+			 * Etc...
+			 */
+			int index = 0;
+			double threshold = 0.5;
+			while (threshold < chance && index < instances.size() - 1) {
+				index++;
+				threshold = Math.pow(0.5, index + 1);
+			}
+
+			return splice(instances, index);
+		}
+	};
+
+	public static final RoutingStrategy ROUND_ROBIN = new RoutingStrategy() {
+
+		private final Multimap<String, ServiceInstance> lastRequested = HashMultimap.create();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> nextAvailable = serviceLocator.listInstances(serviceName);
+			Set<ServiceInstance> lastRequests = Sets.newHashSet(lastRequested.get(serviceName));
+
+			// If all instances have been hit at least once, reset the counter for this service.
+			if (lastRequests.containsAll(nextAvailable)) {
+				lastRequested.removeAll(serviceName);
+			}
+
+			// List the instances to hit in order of not hit to already hit (conserving network distance).
+			List<ServiceInstance> nextUp = nextAvailable.stream()
+					.filter(next -> !lastRequests.contains(next))
+					.collect(Collectors.toList());
+			nextAvailable.stream()
+					.filter(lastRequests::contains)
+					.forEachOrdered(nextUp::add);
+
+			// Register that nextUp[0] instance is going to get hit first.
+			if (!nextUp.isEmpty()) {
+				lastRequested.put(serviceName, nextUp.get(0));
+			}
+
+			return nextUp;
+		}
+	};
```

Comment:
I'm not too sure about this implementation of round robin, because of two things: 

1. It returns a *list* of instances, but it assumes only (and always) the first one is going to be used. This doesn't hold: take for example the `crawlerFilter`, it can dispatch to multiple services simultaneously, and it is not unlikely that other services will implement fallbacks to other services as well. I think in order to address this properly you should either only return a single instance (i.e. replace `listInstances` with `getNextInstance`), or accept a consumer that takes a service instance (i.e. replace `listInstances` with `withInstance(Consumer<ServiceInstance>)`. That would require users of the `RoutingStrategy` to fully rely on the `RoutingStrategy` class for routing, and not implement their own on top of it (like `crawlerFilter` does now).

2. The clean up logic is flawed. It clears the `lastRequested` list when all services has been visited at once, but that means that services can be hit shortly after each other (e.g. the last one in "session" n can be the first one in "session" n+1, meaning it is scheduled twice directly after each other). Since round robin is most useful for "heavy" tasks (cpu/memory/io intensive) this can actually be problematic. A proper implementation should actually keep track of which instances were requested the longest time ago. I would therefore suggest to make `lastRequests` a list, and use that list to sort the `nextAvailable` list in order of last used. Instead of clearing the set you then also only need to move the used instance in the list.